src/main.rs

```diff
@@ -1,100 +1,170 @@
+
+/// Optivanity: hyper-parallelized vanity address generator for the Aptos blockchain, brought to you
+/// by Econia Labs
+#[derive(Parser, Debug)]
+struct CliArgs {
+    /// Address prefix to match (no leading 0x). Each additional character slows search by 16x.
+    #[arg(short, long)]
+    prefix: String,
+    /// Use this flag if you want to search for multisig address(es)
+    #[arg(short, long)]
+    multisig: bool,
+    /// Number of vanity accounts to generate
+    #[arg(short, long, default_value_t = 1)]
+    count: u64,
+    /// Number of threads to use. Only specify if you want to use fewer cores than available.
+    #[arg(short, long, default_value_t = available_parallelism().unwrap().get())]
+    threads: usize,
+}
 
 /// Derive authentication key bytes vector from a reference to a private key.
 fn auth_key_bytes_vec(private_key_ref: &Ed25519PrivateKey) -> Vec<u8> {
     AuthenticationKey::ed25519(&Ed25519PublicKey::from(private_key_ref)).to_vec()
 }
 
-fn main() {
-    // Get command arguments.
-    let args: Vec<_> = env::args().collect();
-    let mut check_str = args[PREFIX_ARG].clone(); // Get prefix string for format check.
-    let is_odd_length = check_str.len() % 2 != 0; // Check if string has odd character count.
-    if is_odd_length {
-        // Append a 0 to check string if an odd number of hex characters.
-        check_str.push('0')
+/// Parse command line arguments, verifying hex characters and specified thread count.
+fn parse_args() -> Result<CliArgs> {
+    let args = CliArgs::parse();
+
+    // Verify prefix has valid hex characters, appending a 0 as needed for an odd character count.
+    if has_odd_character_count(&args.prefix) {
+        hex::decode(format!("{}0", &args.prefix))?;
+    } else {
+        hex::decode(&args.prefix)?;
+    }
+
+    Ok(args)
+}
+
+/// Return `true` if string reference indicates a string with an odd number of characters.
+fn has_odd_character_count(string_ref: &str) -> bool {
+    string_ref.len() % 2 == 1
+}
+
+/// Generate a private key corresponding to a vanity prefix, while search is ongoing.
+///
+/// Once a match is found, a match message is transmitted to the main thread. Once the main thread
+/// has received sufficient match transactions, it broadcasts an exit transaction to search threads.
+///
+/// # Arguments
+///
+/// * `prefix` - The vanity prefix to search against
+/// * `multisig` - If `true` search for a multisig address
+/// * `match_tx` - Transmit channel for match message sent to main thread when a match is found
+/// * `exit_rx` - Receive channel for exit message broadcast from main thread when search is done
+fn generate_key(
+    prefix: String,
+    multisig: bool,
+    match_tx: channel::Sender<()>,
+    exit_rx: channel::Receiver<()>,
+) -> Result<()> {
+    // Clone prefix for getting bytes to check against.
+    let mut check_str = prefix.clone();
+    // Pop last character of check string if an odd length.
+    if has_odd_character_count(&check_str) {
```

Comment:
I think this is a bug. You turn an odd character count prefix into an even one by adding 0, and then pop off a character only for odd prefixes. So this branch will never get hit because you're only ever passing even-length prefixes, causing a 16x slow-down for odd prefixes (due to the zero being retained). Correct me if I'm wrong!

Reply:
> You turn an odd character count prefix into an even one by adding 0, and then pop off a character only for odd prefixes

Note that the check where a 0 is appended does not mutate the user-specified prefix in the case that is has an odd length. Hence the prefix that gets passed in here may have an odd number of characters, but gets optionally truncated to an even number via `check_str`